src/Mono.Android/Java.Interop/TypeManager.cs

```diff
@@ -9,48 +9,54 @@
 
 namespace Java.Interop {
 
-	public static partial class TypeManager {
-
-		// Make this internal so that JNIEnv.Initialize can trigger the static
-		// constructor so that JNIEnv.RegisterJNINatives() doesn't include 
-		// the static constructor execution.
-
-		// Lock on jniToManaged before accessing EITHER jniToManaged or managedToJni.
-		internal static Dictionary<string, Type> jniToManaged = new Dictionary<string, Type> ();
-		static Dictionary<Type, string> managedToJni = new Dictionary<Type, string> ();
-
-		internal static IntPtr id_Class_getName;
-
-		static TypeManager ()
-		{
-			var start = new DateTime ();
-			if (Logger.LogTiming) {
-				start = DateTime.UtcNow;
-				Logger.Log (LogLevel.Info,
-						"monodroid-timing",
-						"TypeManager.cctor start: " + (start - new DateTime (1970, 1, 1)).TotalMilliseconds);
+	static class TypeManagerMapDictionaries
+	{
+		static Dictionary<string, Type> _jniToManaged;
+		static Dictionary<Type, string> _managedToJni;
+
+		public static readonly object AccessLock = new object ();
+
+		//
+		// Access to both properties MUST be done after taking lock on accessLock!
+		//
+		public static Dictionary<string, Type> JniToManaged {
+			get {
+				if (_jniToManaged == null)
+					_jniToManaged = new Dictionary<string, Type> ();
+				return _jniToManaged;
 			}
+		}
 
-			__TypeRegistrations.RegisterPackages ();
-
-			if (Logger.LogTiming) {
-				var end = DateTime.UtcNow;
-				Logger.Log (LogLevel.Info,
-						"monodroid-timing",
-						"TypeManager.cctor time: " + (end - new DateTime (1970, 1, 1)).TotalMilliseconds + " [elapsed: " + (end - start).TotalMilliseconds + " ms]");
+		public static Dictionary<Type, string> ManagedToJni {
+			get {
+				if (_managedToJni == null)
+					_managedToJni = new Dictionary<Type, string> ();
+				return _managedToJni;
 			}
 		}
+	}
+
+	public static partial class TypeManager {
+		[DllImport ("__Internal", CallingConvention = CallingConvention.Cdecl)]
+		extern static IntPtr monodroid_TypeManager_get_java_class_name (IntPtr klass);
+
+		[DllImport ("__Internal", CallingConvention = CallingConvention.Cdecl)]
+		extern static void monodroid_free (IntPtr ptr);
```

Comment:
`AndroidEnvironment` already has this P/Invoke declaration.  We should make the `AndroidEnvironment` method internal, or move it to some shared location, instead of having two separate `monodroid_free()` declarations.